### PR TITLE
Implement SourceForge checkver functionality

### PR DESCRIPTION
**Pull request by se35710**
#### Description
In it's most simple form:
```
    "checkver": "sourceforge",
```
This will find the project name from homepage, or if that fails, from the manifest name.

A more comples example:
```
    "checkver": {
        "sourceforge": {
            "project": "pmt",
            "path": "/pngcrush-executables"
        }
    },
```
Here we provide the project name, and a path into the files section (to limit number of downloads). `project` is optional and only needed if `homepage` or manifest name is not equal to the SourceForge project name.

Regex can be applied as normal, on the `checkver` property, or on the `re` property. The default regex is `"(?:.*)?\/(?:v)?([\d.]+)\/"` plus the path to the project.

#### Motivation and Context
We have a lot of different approaches to SourceForge updates, this will help with maintaining those.

Duplicate of #2455 

#### How Has This Been Tested?
I have updated manifests in the original PR (#2455), which I have run `checkver -u -f` on.

#### Checklist:

- [X] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

File: bin/checkver.ps1 [hunk 1]
```diff
@@ -132,7 +133,35 @@ $Queue | ForEach-Object {
...
+    if ($json.checkver.sourceforge.project) {
+        $url = 'https://sourceforge.net/projects/' + $json.checkver.sourceforge.project + '/rss'
+        $regex = '\/\/sourceforge\.net\/projects\/' + $json.checkver.sourceforge.project + $sourceforgeRegex
+    }
+
+    if ($json.checkver.sourceforge.path) {
+        $url = $url + '?path=' + $json.checkver.sourceforge.path
+    }
+
+    if($json.checkver.re) {
```
Comment (rashil2000) [inline bin/checkver.ps1:164]:
```suggestion
    if ($json.checkver.re) {
```

(format) space

**Comment by rashil2000** [comment]
Do you have any example manifests to test this against?

File: bin/checkver.ps1 (hunk 1 above)
Comment (se35710) [reply bin/checkver.ps1:164]: Oh, yes. Bad picking in the merge. Thanks for finding it! Fix coming right up!

**Comment by se35710** [comment]
Yes, in the original PR (#2455). Since then the bucket was moved from Scoop to Main, so I removed them all from this PR.

**Comment by se35710** [comment]
@rashil2000 Is this ready for push to develop?

**Comment by niheaven** [comment]
Could you please merge 'develop' and add a CHANGELOG item? Put this PR under `Features` before `Bug Fixes`.

File: bin/checkver.ps1 [hunk 2]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
...
 
+    if ($json.checkver -eq 'sourceforge' -or $json.homepage.Contains('sourceforge.net') -or $json.checkver.sourceforge.path) {
+        if ($json.homepage -match '\/\/sourceforge\.net\/projects\/(?<project>([\w-]+$|[\w-]+))(?:[/]?)' -or $json.homepage -match '\/\/(?<project>[\w-]+)\.sourceforge\.net') {
+            $url = 'https://sourceforge.net/projects/' + $matches['project'] + '/rss'
+            $regex = '\/\/sourceforge\.net\/projects\/' + $matches['project'] + $sourceforgeRegex
+            if ($json.checkver -ne 'sourceforge' -and $json.checkver.GetType() -eq [System.String]) {
+                $regex = $json.checkver
+            }
+        }
+        else {
```
Comment (rashil2000) [inline bin/checkver.ps1:144]:
```suggestion
        } else {
```

**Review by rashil2000 — APPROVED** [review]

**Comment by rashil2000** [comment]
I tested with 6 manifests - xmlstarlet, potrace, pngcrush, editorconfig, graphicsmagick, bochs

- graphicsmagick (with this PR) reports version better than current checkver (without this PR)
- editorconfig checkver does not work with this PR
- rest 4 manifests work as expected, i.e. the same as current checkver

File: schema.json [hunk 3]
```diff
@@ -306,6 +306,25 @@
...
+                                {
+                                    "additionalProperties": false,
+                                    "properties": {
+                                        "project": {
+                                            "type": "string"
+                                        },
+                                        "path": {
+                                            "type": "string"
+                                        }
+                                    }
```
Comment (niheaven) [inline schema.json:325]:
```suggestion
                                    },
                                    "type": "object"
```

File: bin/checkver.ps1 [hunk 4]
```diff
@@ -106,6 +106,7 @@ $Queue | ForEach-Object {
     Register-ObjectEvent $wc downloadstringcompleted -ErrorAction Stop | Out-Null
 
     $githubRegex = '\/releases\/tag\/(?:v|V)?([\d.]+)'
+    $sourceforgeRegex = '(?:.*)?\/(?:v)?([\d.]+)\/'
```
Comment (niheaven) [inline bin/checkver.ps1:109]:
```suggestion
    $githubRegex = '/releases/tag/(?:v|V)?([\d.]+)'
    $sourceforgeRegex = '(?:.*)?/(?:v)?([\d.]+)/'
```

File: bin/checkver.ps1 [hunk 5]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
         if ($json.checkver.PSObject.Properties.Count -eq 1) { $useGithubAPI = $true }
     }
 
+    if ($json.checkver -eq 'sourceforge' -or $json.homepage.Contains('sourceforge.net') -or $json.checkver.sourceforge.path) {
+        if ($json.homepage -match '\/\/sourceforge\.net\/projects\/(?<project>([\w-]+$|[\w-]+))(?:[/]?)' -or $json.homepage -match '\/\/(?<project>[\w-]+)\.sourceforge\.net') {
```
Comment (niheaven) [inline bin/checkver.ps1:137]:
```suggestion
        if ($json.homepage -match '//sourceforge\.net/projects/(?<project>([\w-]+$|[\w-]+))(?:[/]?)' -or $json.homepage -match '//(?<project>[\w-]+)\.sourceforge\.net') {
```

File: bin/checkver.ps1 [hunk 6]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
...
+            $url = 'https://sourceforge.net/projects/' + $matches['project'] + '/rss'
+            $regex = '\/\/sourceforge\.net\/projects\/' + $matches['project'] + $sourceforgeRegex
```
Comment (niheaven) [inline bin/checkver.ps1:139]:
```suggestion
            $regex = '//sourceforge\.net/projects/' + $matches['project'] + $sourceforgeRegex
```

File: bin/checkver.ps1 [hunk 7]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
...
+            if ($json.checkver -ne 'sourceforge' -and $json.checkver.GetType() -eq [System.String]) {
+                $regex = $json.checkver
+            }
+        }
+        else {
+            $url = 'https://sourceforge.net/projects/' + (strip_ext $name) + '/rss'
+            $regex = '\/\/sourceforge\.net\/projects\/' + (strip_ext $name) + $sourceforgeRegex
```
Comment (niheaven) [inline bin/checkver.ps1:146]:
```suggestion
            $regex = '//sourceforge\.net/projects/' + (strip_ext $name) + $sourceforgeRegex
```

File: bin/checkver.ps1 [hunk 8]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
...
+        }
+    }
+
+    if ($json.checkver.sourceforge -and $json.checkver.sourceforge.GetType() -eq [System.String]) {
+        $url = 'https://sourceforge.net/projects/' + $json.checkver.sourceforge + '/rss'
+        $regex = '\/\/sourceforge\.net\/projects\/' + $json.checkver.sourceforge + $sourceforgeRegex
```
Comment (niheaven) [inline bin/checkver.ps1:152]:
```suggestion
        $regex = '//sourceforge\.net/projects/' + $json.checkver.sourceforge + $sourceforgeRegex
```

File: bin/checkver.ps1 [hunk 9]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
...
+    }
+
+    if ($json.checkver.sourceforge.project) {
+        $url = 'https://sourceforge.net/projects/' + $json.checkver.sourceforge.project + '/rss'
+        $regex = '\/\/sourceforge\.net\/projects\/' + $json.checkver.sourceforge.project + $sourceforgeRegex
```
Comment (niheaven) [inline bin/checkver.ps1:157]:
```suggestion
        $regex = '//sourceforge\.net/projects/' + $json.checkver.sourceforge.project + $sourceforgeRegex
```

File: bin/checkver.ps1 [hunk 10]
```diff
@@ -132,6 +133,34 @@ $Queue | ForEach-Object {
...
+    }
+
+    if ($json.checkver.sourceforge.path) {
+        $url = $url + '?path=' + $json.checkver.sourceforge.path
```
Comment (niheaven) [inline bin/checkver.ps1:161]:
```suggestion
        $url = $url + '?path=/' + $json.checkver.sourceforge.path.TrimStart('/')
```

**Review by niheaven — COMMENTED** [review]
- No need to escape `/` in PowerShell regex, ref https://docs.microsoft.com/en-us/dotnet/standard/base-types/character-escapes-in-regular-expressions#character-escapes-in-net
- Add `"type": "object"` to schema
- For apps such as `editorconfig` that have sub-path, maybe a `subpath` sec is preferred

**Comment by se35710** [comment]
I'll put this on hold until I can figure out what my rebase did,

**Comment by mikefarmer01** [comment]
@se35710 Any updates, dear Mathias?

**Comment by niheaven** [comment]
@se35710 I've kept a copy since last review, and could I submit that copy and make some tweaks?

**Comment by se35710** [comment]
@niheaven That would be awesome. Please do!

**Pull request closed without merging**